src/storage/mvcc/reader/reader.rs

```diff
@@ -1,77 +1,78 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
 use crate::storage::kv::{Cursor, CursorBuilder, ScanMode, Snapshot, Statistics};
-use crate::storage::mvcc::{default_not_found_error, Result};
-use engine_traits::{IterOptions, MvccProperties};
+use crate::storage::mvcc::{
+    default_not_found_error,
+    reader::{OverlappedWrite, TxnCommitRecord},
+    seek_for_valid_write, Result,
+};
+use engine_traits::IterOptions;
 use engine_traits::{CF_DEFAULT, CF_LOCK, CF_WRITE};
-use kvproto::kvrpcpb::IsolationLevel;
-use std::borrow::Cow;
-use txn_types::{Key, Lock, TimeStamp, Value, Write, WriteRef, WriteType};
-
-const GC_MAX_ROW_VERSIONS_THRESHOLD: u64 = 100;
-
-/// The result of `get_txn_commit_record`, which is used to get the status of a specified
-/// transaction from write cf.
-#[derive(Debug)]
-pub enum TxnCommitRecord {
-    /// The commit record of the given transaction is not found. But it's possible that there's
-    /// another transaction's commit record, whose `commit_ts` equals to the current transaction's
-    /// `start_ts`. That kind of record will be returned via the `overlapped_write` field.
-    /// In this case, if the current transaction is to be rolled back, the `overlapped_write` must not
-    /// be overwritten.
-    None {
-        overlapped_write: Option<OverlappedWrite>,
-    },
-    /// Found the transaction's write record.
-    SingleRecord { commit_ts: TimeStamp, write: Write },
-    /// The transaction's status is found in another transaction's record's `overlapped_rollback`
-    /// field. This may happen when the current transaction's `start_ts` is the same as the
-    /// `commit_ts` of another transaction on this key.
-    OverlappedRollback { commit_ts: TimeStamp },
+use txn_types::{Key, Lock, OldValue, TimeStamp, Value, Write, WriteRef, WriteType};
+
+/// Read from an MVCC snapshot, i.e., a logical view of the database at a specific timestamp (the
+/// start_ts).
+///
+/// This represents the view of the database from a single transaction.
+///
+/// Confusingly, there are two meanings of the word 'snapshot' here. In the name of the struct,
```

Comment:
How about make `SnapshotReader` into `MvccSnapshotReader` to make it clear.
Or we can also make `Snapshot` into `EngineSnapshot`

Reply:
I made the second change. I feel that the `MvccSnapshotReader` us a bit long